### ISSUE TITLE
Error when analyzing the native code in (*.so form) from Android apps

### ISSUE THREAD

**Issue by 040840308** [cea-sec/miasm#765]
I tried to load a so file into IDA 7.0 and launched the script depgraph.py.
However, I found an error as follows:

C:\Program Files\IDA 7.0\python\depgraph.py: 
Traceback (most recent call last):
  File "C:\Program Files\IDA 7.0\python\ida_idaapi.py", line 553, in IDAPython_ExecScript
    execfile(script, g)
  File "C:/Program Files/IDA 7.0/python/depgraph.py", line 277, in <module>
    launch_depgraph()
  File "C:/Program Files/IDA 7.0/python/depgraph.py", line 238, in launch_depgraph
    settings = depGraphSettingsForm(ir_arch)
  File "C:/Program Files/IDA 7.0/python/depgraph.py", line 38, in __init__
    assert cur_block is not None
AssertionError

**Comment by serpilliere** [comment]
Hi!
Can you provide more informations?
- miasm revision
- target architecture, 
- ...


By the way: it may not be related, but there was an api fix in pr #766

**Comment by 040840308** [comment]
I installed miasm2 on Win10, according to the instructions:
pip install pyparsing elfesteem
Copy miasm2/miasm2 directory to C:\...\IDA\python\, the most recent version on the github
target architecture is 'arml', the native code from an APK.

**Comment by 040840308** [comment]
By the way, I also tried your mentioned an api fix in pr #766. However, it does not work.
The error is caused by the code in function depGraphSettingsFrom,  the code " ira.getby_offset(self.address)" returns an empty set. as follows:

class depGraphSettingsForm(ida_kernwin.Form):
    def __init__(self, ira):
        self.ira = ira
        self.stk_args = {'ARG%d' % i:i for i in xrange(10)}
        self.stk_unalias_force = False
        self.address = idc.ScreenEA()
        cur_block = None
        print ira.getby_offset(self.address)
        for block in ira.getby_offset(self.address):
            offset = self.ira.symbol_pool.loc_key_to_offset(block.loc_key)
            print offset
            if offset is not None:
                # Only one block non-generated
                assert cur_block is None
                cur_block = block
        **assert cur_block is not None**

Why this happens?

**Comment by serpilliere** [comment]
Just a side note:
you have the wrong elfesteem. As the documentation says, you have to install the version from github/serpilliere

Can you provide the binary or an example that mimic the behaviour?

Thanks!

**Comment by 040840308** [comment]
YES, here is my test examples: https://github.com/040840308/ExampelForMiasm.git. I have reinstalled elfesteem and try again. However, it still does not work. 
Does miasm work on Win10?
I used the following installation instructions:
git clone https://github.com/serpilliere/elfesteem.git elfesteem
cd elfesteem
python setup.py build
python setup.py install
pip install pyparsing
Copy miasm2/miasm2 directory to C:\...\IDA\python\
Copy elfesteem directory and pyparsing.py to C:\...\IDA\python\

**Comment by commial** [comment]
Hi,

Windows is not officially supported by Miasm. But, as stated in #629,  you can try using the Windows Sub-Linux system to have it running.

Actually, any part related to the use of the virtual memory (VmMngr, emulation and so one) will not work.
But the others parts still have a chance to work, because no installation / compilation is required.

Also, thanks for providing us an example. Could you also provide the targeted element for DepGraph use?

**Comment by 040840308** [comment]
OK, I just want to combine IDA with miasm on Win10, according to the documentation.

Windows & IDA
Most of Miasm's IDA plugins use a subset of Miasm functionnality. A quick way to have them working is to add:
elfesteem directory and pyparsing.py to C:\...\IDA\python\ or pip install pyparsing elfesteem
miasm2/miasm2 directory to C:\...\IDA\python\
All features excepting JITter related ones will be available. For a more complete installation, please refer to above paragraphs.

**Comment by serpilliere** [comment]
Ok, I see the problem:
Your library is mixed between ARM and ARMTHUMB and we don't kknow how to extract the information from IDA.
Here, the code disasm in ARM which doesn't produce blocks, which makes the lookup fail.


Does someone here know how to distinguish ARM from ARMT in IDA?

**Comment by 040840308** [comment]
OK. Thus, if we miasm can distinguish ARM from ARMT in IDA, can this problem be solved?

**Comment by serpilliere** [comment]
Yes, I think so.
To test this you can fix the actual code to force armt. In the file, replace:
```python
        is_armt = globals().get('armt', False)
```
with:
```python
        is_armt = True
```


(yes, the original code test a global, but this code in wrong in IDA)

Tell me if this works

**Comment by 040840308** [comment]
Nice, it does work currently. (ps. I can run degraph.py on Win10 with IDA 7.0. Thus, other who wants to this can try in this way)
Later, I will try to make a large scale analysis using miasm.  I will tell you if have other questions.
Thanks a lot.

**Comment by 040840308** [comment]
By the way, are there some APIs to iterate the generated DependencyGraph? For example, extract the nodes from the DependencyGraph? I searched the online document, but didn't find it.

**Comment by serpilliere** [comment]
Fine!
So the remaining problem is how to find the way in IDA to distinguish ARM from ARMT :smile: 

For your question, look at the example:
https://github.com/cea-sec/miasm/blob/master/example/symbol_exec/depgraph.py

**Comment by 040840308** [comment]
OK. Do you mean that I still need to do some work to distinguish ARM from ARMT in IDA?

**Comment by serpilliere** [comment]
No no: the test has to be added in Miasm :smiley:

**Comment by 040840308** [comment]
OK, I'm new to Miasm and wish someone can solve this.

**Comment by serpilliere** [comment]
I'm on it !

**Comment by 040840308** [comment]
Nice.
By the way, I looked your mentioned example: https://github.com/cea-sec/miasm/blob/master/example/symbol_exec/depgraph.py. However, I only find this from this example, as follows:
# Get the instance
dg = DependencyGraph(ir_arch, implicit=args.implicit, apply_simp=not args.do_not_simplify, follow_mem=not args.unfollow_mem, follow_call=not args.unfollow_call)
# Build information
target_addr = int(args.target_addr, 0)
current_block = list(ir_arch.getby_offset(target_addr))[0]
assignblk_index = 0
for assignblk_index, assignblk in enumerate(current_block):
    if assignblk.instr.offset == target_addr:
        break
# Enumerate solutions
json_solutions = []
for sol_nb, sol in enumerate(dg.get(current_block.loc_key, elements, assignblk_index, set())):
    fname = "sol_%d.dot" % sol_nb
    with open(fname, "w") as fdesc:
            fdesc.write(sol.graph.dot())
    results = sol.emul(ctx=init_ctx)
    tokens = {str(k): str(v) for k, v in results.iteritems()}
    if not args.json:
        result = ", ".join("=".join(x) for x in tokens.iteritems())
        print "Solution %d: %s -> %s" % (sol_nb, result, fname)
        if sol.has_loop:
            print '\tLoop involved'

Could I extract DependencyNodes or  DependencyResult from the generated DependencyGraph directly?

**Comment by commial** [comment]
Hi,

I'm not sure to fully understand your question, but:
With an `IntermediateRepresentation` object (usually `ir`, `ira`), you obtain a `DependencyGraph` object through `dg = DependencyGraph(ir_arch)`.

You can then ask for multiple session of dependency graph use, through the API: `dg.get`. It will returns an iterator on `DependencyResult`, which contains:
* a graph representing the solution (in which edges are dependencies and nodes are `DependencyNodes`)
* API to get back the values of what you asked: `results = sol.emul()` (ie. "give me the final value of the tracked register RAX")

Is it more clear?

**Comment by 040840308** [comment]
Thanks. Maybe, I should read more details about Miasm seriously.

**Comment by serpilliere** [comment]
The ARMT guess should be resolved by #770

**Comment by commial** [comment]
As the bug has been fixed and it seems to work, I'm closing the issue.
Please open a new one if you have more question on a different topic.

Again, thanks for reporting.
